code/game/objects/items/reagent_containers/hypospray.dm

```diff
@@ -16,72 +16,286 @@
+		return
+	if(isturf(user.loc))
+		to_chat(user, "<span class='notice'>You flush the contents of [src].</span>")
+		reagents.reaction(user.loc)
+		reagents.clear_reagents()
+
+/obj/item/reagent_container/hypospray/proc/label(mob/user)
+	var/str = copytext(reject_bad_text(input(user,"Label text?", "Set label", "")), 1, MAX_NAME_LEN)
+	if(!str || !length(str))
+		to_chat(user, "<span class='notice'>Invalid text.</span>")
+		return
+	to_chat(user, "<span class='notice'>You label [src] as \"[str]\".</span>")
+	name = "[core_name] ([str])"
+	label = str
 
 /obj/item/reagent_container/hypospray/attack_paw(mob/user as mob)
 	return attack_hand(user)
 
-/obj/item/reagent_container/hypospray/attack(mob/M, mob/living/user)
+/obj/item/reagent_container/hypospray/afterattack(obj/T, mob/living/user)
+	if(!T.reagents)
+		return
+	if(!inject_mode) //if we're draining
+		if(reagents.holder_full())
+			to_chat(user, "<span class='warning'>[src] is full.</span>")
+			return
+
+		if(ismob(T))//Blood!
+			if(iscarbon(T))
+				var/amount = min(src.reagents.maximum_volume - src.reagents.total_volume, amount_per_transfer_from_this)
+				var/mob/living/carbon/C = T
+				if(C.get_blood_id() && reagents.has_reagent(C.get_blood_id()))
+					to_chat(user, "<span class='warning'>There is already a blood sample in [src].</span>")
+					return
+				if(!C.dna)
+					to_chat(user, "<span class='warning'>You are unable to locate any blood.</span>")
+					return
+				if(NOCLONE in C.mutations) //target done been et, no more blood in him
+					to_chat(user, "<span class='warning'>You are unable to locate any blood.</span>")
+					return
+
+				if(ishuman(T))
+					var/mob/living/carbon/human/H = C
+					if(H.species.flags & NO_BLOOD)
+						to_chat(user, "<span class='warning'>You are unable to locate any blood.</span>")
+						return
+					else
+						C.take_blood(src,amount)
+				else
+					C.take_blood(src,amount)
+
+				on_reagent_change()
+				reagents.handle_reactions()
+				user.visible_message("<span clas='warning'>[user] takes a blood sample from [T].</span>",
+									"<span class='notice'>You take a blood sample from [T].</span>", null, 4)
+		else //if not mob
+			if(!T.reagents.total_volume)
+				to_chat(user, "<span class='warning'>[T] is empty.")
+				return
+
+			if(!T.is_drawable())
+				to_chat(user, "<span class='warning'>You cannot directly remove reagents from this object.</span>")
+				return
+
+			var/trans = T.reagents.trans_to(src, amount_per_transfer_from_this)
+
+			to_chat(user, "<span class='notice'>You fill [src] with [trans] units of the solution.</span>")
+		if (reagents.holder_full())
+			inject_mode = !inject_mode
+			update_icon()
+		return TRUE
+
+
 	if(!reagents.total_volume)
-		to_chat(user, "<span class ='danger'>[src] is empty!</span>")
+		to_chat(user, "<span class='warning'>[src] is empty.</span>")
 		return
-	if (!istype(M))
+	if(!T.is_injectable() && !ismob(T))
```

Comment:
```suggestion
	if(!A.is_injectable())
```

Reply:
you dont need to check for mobs, is_injectable() already covers this.

Reply:
`is_injectable(FALSE)` to be specific, it has a var called `allowmobs` for it.